internals-js/src/federation.ts

```diff
@@ -641,52 +641,66 @@ export class FederationMetadata {
+  // we include @shareable or @override in those "legacy" directives).
+  private getLegacyFederationDirective<TApplicationArgs extends {[key: string]: any}>(
     name: FederationDirectiveName
   ): DirectiveDefinition<TApplicationArgs> {
-    const directive = this.schema.directive(this.federationDirectiveNameInSchema(name));
+    const directive = this.getFederationDirective<TApplicationArgs>(name);
     assert(directive, `The provided schema does not have federation directive @${name}`);
-    return directive as DirectiveDefinition<TApplicationArgs>;
+    return directive;
+  }
+
+  private getFederationDirective<TApplicationArgs extends {[key: string]: any}>(
+    name: FederationDirectiveName
+  ): DirectiveDefinition<TApplicationArgs> | undefined {
+    return this.schema.directive(this.federationDirectiveNameInSchema(name)) as DirectiveDefinition<TApplicationArgs> | undefined;
   }
 
   keyDirective(): DirectiveDefinition<{fields: any, resolvable?: boolean}> {
-    return this.getFederationDirective(FederationDirectiveName.KEY);
+    return this.getLegacyFederationDirective(FederationDirectiveName.KEY);
   }
 
   overrideDirective(): DirectiveDefinition<{from: string}> {
-    return this.getFederationDirective(FederationDirectiveName.OVERRIDE);
+    return this.getLegacyFederationDirective(FederationDirectiveName.OVERRIDE);
   }
 
   extendsDirective(): DirectiveDefinition<Record<string, never>> {
-    return this.getFederationDirective(FederationDirectiveName.EXTENDS);
+    return this.getLegacyFederationDirective(FederationDirectiveName.EXTENDS);
   }
 
   externalDirective(): DirectiveDefinition<{reason: string}> {
-    return this.getFederationDirective(FederationDirectiveName.EXTERNAL);
+    return this.getLegacyFederationDirective(FederationDirectiveName.EXTERNAL);
   }
 
   requiresDirective(): DirectiveDefinition<{fields: any}> {
-    return this.getFederationDirective(FederationDirectiveName.REQUIRES);
+    return this.getLegacyFederationDirective(FederationDirectiveName.REQUIRES);
   }
 
   providesDirective(): DirectiveDefinition<{fields: any}> {
-    return this.getFederationDirective(FederationDirectiveName.PROVIDES);
+    return this.getLegacyFederationDirective(FederationDirectiveName.PROVIDES);
   }
 
   shareableDirective(): DirectiveDefinition<{}> {
-    return this.getFederationDirective(FederationDirectiveName.SHAREABLE);
+    return this.getLegacyFederationDirective(FederationDirectiveName.SHAREABLE);
   }
 
   tagDirective(): DirectiveDefinition<{name: string}> {
-    return this.getFederationDirective(FederationDirectiveName.TAG);
+    return this.getLegacyFederationDirective(FederationDirectiveName.TAG);
   }
 
-  composeDirective(): DirectiveDefinition<{name: string}> {
-    return this.getFederationDirective(FederationDirectiveName.COMPOSE_DIRECTIVE);
+  composeDirective(): DirectiveDefinition<{name: string}> | FederationDirectiveNotDefinedInSchema<{name: string}> {
```

Comment:
That's the bit of "magic" needed so that new releases of `internals-js` continue to work with older `@apollo/composition`, because `composeDirectiveManager` calls
```typescript
const composeDirectives = sg.metadata()
    .composeDirective()
    .applications();
```
and that would start failing if we made `FederationMetadata.composeDirective()` sometimes return `undefined`.

Note that I tried to make this as safe as possible through typing, and I think that's an ok solution, but returning `undefined` would probably be simpler overall, and we can change to that later if we fix-up our locking of versions between packages.